Fix #976: Add support for websocket send_json and receive_json

vitaliemaldur opened the pull request:
<!-- Thank you for your contribution!

     Before submit your Pull Request, make sure you picked
     the right branch:

     - If you propose a new feature or improvement, select "master"
       as a target branch;
     - If this is a bug fix or documentation amendment, select
       the latest release branch (which looks like "0.xx") -->

## What do these changes do?

Add support for websocket server to send json and for websocket client to send and receive json

<!-- Please give a short brief about these changes. -->

## Are there changes in behavior for the user?
User can use methods:
`send_json` and `receive_json` both on websocket server and client
<!-- Outline any notable behaviour for the end users. -->

## Related issue number
#976 
<!-- Are there any issues opened that will be resolved by merging this change? -->

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes

asvetlov commented on aiohttp/web_ws.py:
```diff
@@ -154,6 +154,13 @@ def send_bytes(self, data):
                             type(data))
         self._writer.send(data, binary=True)
 
+    def send_json(self, data, *, dumps=json.dumps):
```
Please call `.send_text()` instead of working on low level.

asvetlov commented on aiohttp/websocket_client.py:
```diff
@@ -88,6 +89,11 @@ def send_bytes(self, data):
         self._writer.send(data, binary=True)
 
+    def send_json(self, data, *, dumps=json.dumps):
+        if self._closed:
```
Use `send_str()`

asvetlov commented:
Would you add tests also?

vitaliemaldur commented:
Yes I would

vitaliemaldur commented:
@asvetlov what I can do to avoid
```
AttributeError: 'module' object has no attribute 'JSONDecodeError'
tests\test_web_websocket.py:161: AttributeError
```
I am using `json.decoder.JSONDecodeError` in a `try/except`.

Also, do you have some suggestions for the tests I wrote?

asvetlov commented on aiohttp/web_ws.py:
```diff
@@ -154,6 +154,13 @@ def send_bytes(self, data):
         self._writer.send(data, binary=True)
 
+    def send_json(self, data, *, dumps=json.dumps):
+        if self._writer is None:
```
These checks are not required: `.send_str()` performs them anyway.
There is no need to do the check twice.

asvetlov commented on tests/test_web_websocket.py:
```diff
@@ -128,6 +144,25 @@ def receive():
+            yield from ws.prepare(req)
+
+            @asyncio.coroutine
+            def receive():
```
Use functional test for the case.

vitaliemaldur commented:
@asvetlov I've noticed that in `web_ws.py` we have `send_str`, `send_bytes`, `send_json` methods and coresponding `receive_str`, `receive_bytes`, `receive_json`. But in `websocket_client` we have `send_str`, `send_bytes`, `send_json` methods and only `receive_json`. Is that on purpose? or should I add the rest of methods there?

asvetlov commented:
I think yes. 
From my perspective we should support the same methods for both client and server for consistency.

vitaliemaldur commented:
@asvetlov I have added all the methods discussed above, now we have:
- `send_str`
- `send_bytes`
- `send_json`
- `receive_str`
- `receive_bytes`
- `receive_json`

both on `websocket_client.py` and `web_ws.py`. I have also added documentation for them.

asvetlov commented:
Thanks!
